[dynamo] Add more dynamo call_methods and getattr support or Placement

yoyoyocmu opened the pull request:
Summary:
Explained by title.
This fix is part of: https://github.com/pytorch/pytorch/issues/117670

Test Plan:
Unit tetst and CI
- Unit test: `buck2 test mode/dev-nosan //caffe2/test/distributed/_tensor:dtensor_compile -- test_placement_compile`

Differential Revision: D52863073




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

wconstab commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -98,6 +99,26 @@ def device_type(self) -> str:
+            a = 0
+            if x.is_replicate():
+                a += 1
+            if x.is_shard():
```
could you also test is_shard(dim) with arg and with kwarg provided?

wconstab commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -106,9 +111,15 @@ def call_method(
-        allowed_methods = ["__init__", "__setattr__"]
-        # placement types dynamo tracking allows only __init__
-        # and __setattr__ methods, the latter is for case like `Shard(dim)`
+        allowed_methods = [
```
it would be better imo to explicitly name this allowed_constant_methods and add a comment explaining that all the methods in this list must satisfy
 - their input arguments are constants and do not need to be guarded on
 - their output is constant with respect to their inputs

bc it would be a risk if someone added a new method to Placement and added it to this list, but it did not meet those requirements

yanboliang commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -112,10 +117,20 @@ def call_method(
+        # Methods in the list must satisfy:
+        #    1. Input arguments are constants and do not need to be guarded on;
+        #    2. Output is constant with respect to their inputs
+        allowed_constant_methods = [
```
Usually we name them as ```constant_fold_functions```.
https://github.com/pytorch/pytorch/blob/c4eab49ded3ebd3742d9d1380f1ef15a312c834d/torch/_dynamo/variables/torch.py#L129

yanboliang commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -112,10 +117,20 @@ def call_method(
+        #    1. Input arguments are constants and do not need to be guarded on;
+        #    2. Output is constant with respect to their inputs
+        allowed_constant_methods = [
+            "__init__",
```
Can you add one example of tracing ```__init__```, I think we should review this carefully since it's probably triggers side effect. In your unit tests, the ```Shard(0)``` and  ```Replicate``` are generated outside of the compile region, I think it won't trigger the tracing of ```__init__```.

yoyoyocmu committed:
[dynamo] Add more dynamo call_methods and getattr support or Placement (#117733)

Summary:

Explained by title.
This fix is part of: https://github.com/pytorch/pytorch/issues/117670

Test Plan:
Unit tetst and CI
- Unit test: `buck2 test mode/dev-nosan //caffe2/test/distributed/_tensor:dtensor_compile -- test_placement_compile`

Differential Revision: D52863073
